{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Okay here comes some notes"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\"\"\"\n",
      "https://www.biostars.org/p/152517/\n",
      "\n",
      "Example of how to work with Ensembl release 81 GTF files, which:\n",
      "\n",
      "    1) already have genes and transcripts included\n",
      "\n",
      "    2) have unique IDs for genes, transcripts, and exons in the corresponding\n",
      "       \"<featuretype>_id\" attribute\n",
      "\n",
      "    3) do not have unique IDs for CDS, stop_codon, start_codon, UTR.\n",
      "\n",
      "See background info at on database IDs at:\n",
      "    https://pythonhosted.org/gffutils/database-ids.html\n",
      "\n",
      "\n",
      "GTF file from\n",
      "\n",
      "ftp://ftp.ensembl.org/pub/release-81/gtf/mus_musculus/Mus_musculus.GRCm38.81.gtf.gz\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "import gffutils\n",
      "import time\n",
      "import argparse\n",
      "\n",
      "ap = argparse.ArgumentParser()\n",
      "ap.add_argument('input', help='Ensembl Release 81 GTF file')\n",
      "ap.add_argument('database', help='Database to be created')\n",
      "ap.add_argument('--testN', type=int,\n",
      "              help='Run a test using only the first N features, and then '\n",
      "                'print out some example feature IDs and their attributes')\n",
      "ap.add_argument('--force', action='store_true',\n",
      "                help='Overwrite an existing database')\n",
      "args = ap.parse_args()\n",
      "\n",
      "\n",
      "def first_n_features(data, n=5000):\n",
      "    \"\"\"\n",
      "    Useful for testing: only use the first `n` features of source data\n",
      "    \"\"\"\n",
      "    for i, feature in enumerate(gffutils.iterators.DataIterator(data)):\n",
      "        if i > n:\n",
      "            break\n",
      "        yield feature\n",
      "\n",
      "\n",
      "# Note: this function is optional; if you don't want these IDs then comment out \n",
      "# the lines at [1] below\n",
      "def subfeature_handler(f):\n",
      "    \"\"\"\n",
      "    Given a gffutils.Feature object (which does not yet have its ID assigned),\n",
      "    figure out what its ID should be.\n",
      "\n",
      "    This is intended to be used for CDS, UTR, start_codon, and stop_codon\n",
      "    features in the Ensembl release 81 GTF files.  I figured a reasonable\n",
      "    unique ID would consist of the parent transcript and the feature type,\n",
      "    followed by an autoincrementing number.\n",
      "\n",
      "    See https://pythonhosted.org/gffutils/database-ids.html#id-spec for\n",
      "    details and other options.\n",
      "    \"\"\"\n",
      "    return ''.join(\n",
      "        ['autoincrement:',\n",
      "         f.attributes['transcript_id'][0],\n",
      "         '_',\n",
      "         f.featuretype])\n",
      "\n",
      "\n",
      "# gffutils can spend a lot of time trying to decide on a unique ID for each\n",
      "# feature. So we have to give it hints of where to look in the attributes.\n",
      "#\n",
      "# We also tell it to use our subfeature_handler function for featuretypes with\n",
      "# no unique IDs.\n",
      "id_spec = {\n",
      "    'exon': 'exon_id',\n",
      "    'gene': 'gene_id',\n",
      "    'transcript': 'transcript_id',\n",
      "\n",
      "    # [1] These aren't needed for speed, but they do give nicer IDs.\n",
      "    'CDS': [subfeature_handler],\n",
      "    'stop_codon': [subfeature_handler],\n",
      "    'start_codon': [subfeature_handler],\n",
      "    'UTR':  [subfeature_handler],\n",
      "}\n",
      "\n",
      "if args.testN is None:\n",
      "    data = args.input\n",
      "else:\n",
      "    data = first_n_features(args.input, args.testN)\n",
      "\n",
      "t0 = time.time()\n",
      "db = gffutils.create_db(\n",
      "    data,\n",
      "    args.database,\n",
      "\n",
      "    # Since Ensembl GTF files now come with genes and transcripts already in\n",
      "    # the file, we don't want to spend the time to infer them (which we would\n",
      "    # need to do in an on-spec GTF file)\n",
      "    disable_infer_genes=True,\n",
      "    disable_infer_transcripts=True,\n",
      "\n",
      "    # Here's where we provide our custom id spec\n",
      "    id_spec=id_spec,\n",
      "\n",
      "    # \"create_unique\" runs a lot faster than \"merge\"\n",
      "    # See https://pythonhosted.org/gffutils/database-ids.html#merge-strategy\n",
      "    # for details.\n",
      "    merge_strategy='create_unique',\n",
      "    verbose=True,\n",
      "    force=args.force,\n",
      ")\n",
      "t1 = time.time()\n",
      "\n",
      "if args.testN is not None:\n",
      "    for ex in ['CDS', 'start_codon', 'stop_codon', 'UTR']:\n",
      "        c = db.features_of_type(ex).next()\n",
      "        print '-----------------------------'\n",
      "        print 'Example \"{0}\" feature; ID=\"{1}\"'.format(ex, c.id)\n",
      "        print 'Attributes:\\n{0}'.format(c.attributes)\n",
      "\n",
      "print \"\\n\\n{0:.1f}s to create database\".format(t1 - t0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "SystemExit",
       "evalue": "2",
       "output_type": "pyerr",
       "traceback": [
        "An exception has occurred, use %tb to see the full traceback.\n",
        "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "usage: -c [-h] [--testN TESTN] [--force] input database\n",
        "-c: error: unrecognized arguments: -f --IPKernelApp.parent_appname='ipython-notebook' --profile-dir --parent=1\n",
        "To exit: use 'exit', 'quit', or Ctrl-D.\n"
       ]
      }
     ],
     "prompt_number": 1
    }
   ],
   "metadata": {}
  }
 ]
}